boards/shared/Src/Io/Io_SharedSpi.c

```diff
@@ -0,0 +1,114 @@
+#include "Io_SharedSpi.h"
+
+#define RETURN_IF_STATUS_NOT_OK(status, nss_port, nss_pin)        \
+    if ((status) != HAL_OK)                                       \
+    {                                                             \
+        HAL_GPIO_WritePin((nss_port), (nss_pin), GPIO_PIN_RESET); \
+        return (status);                                          \
+    }
+
+struct SharedSpi
+{
+    SPI_HandleTypeDef *hspi;
+    GPIO_TypeDef *     nss_port;
+    uint16_t           nss_pin;
+};
+
+struct SharedSpi *Io_SharedSpi_Create(
+    SPI_HandleTypeDef *hspi,
+    GPIO_TypeDef *     nss_port,
+    uint16_t           nss_pin)
+{
+    assert(hspi != NULL);
+
+    struct SharedSpi *spi = malloc(sizeof(struct SharedSpi));
+    assert(spi != NULL);
+
+    spi->hspi     = hspi;
+    spi->nss_pin  = nss_pin;
+    spi->nss_port = nss_port;
+
+    return spi;
+}
+
+void Io_SharedSpi_SetNssLow(const struct SharedSpi *const spi)
+{
+    HAL_GPIO_WritePin(spi->nss_port, spi->nss_pin, GPIO_PIN_RESET);
+}
+
+void Io_SharedSpi_SetNssHigh(const struct SharedSpi *const spi)
+{
+    HAL_GPIO_WritePin(spi->nss_port, spi->nss_pin, GPIO_PIN_SET);
+}
+
+HAL_StatusTypeDef Io_SharedSpi_TransmitAndReceive(
+    const struct SharedSpi *const spi,
+    uint8_t *                     tx_data,
+    uint16_t                      tx_size,
+    uint8_t *                     rx_data,
+    uint16_t                      rx_size)
+
+{
+    Io_SharedSpi_SetNssLow(spi);
+    RETURN_IF_STATUS_NOT_OK(
+        HAL_SPI_Transmit(spi->hspi, tx_data, tx_size, 100U), spi->nss_port,
+        spi->nss_pin)
+    RETURN_IF_STATUS_NOT_OK(
+        HAL_SPI_Receive(spi->hspi, rx_data, rx_size, 100U), spi->nss_port,
+        spi->nss_pin)
+    Io_SharedSpi_SetNssHigh(spi);
+
+    return HAL_OK;
+}
+
+HAL_StatusTypeDef Io_SharedSpi_Transmit(
+    const struct SharedSpi *const spi,
+    uint8_t *                     tx_data,
+    uint16_t                      tx_size)
+{
+    Io_SharedSpi_SetNssLow(spi);
+    HAL_StatusTypeDef status =
+        HAL_SPI_Transmit(spi->hspi, tx_data, tx_size, 100U);
+    Io_SharedSpi_SetNssHigh(spi);
+
+    return status;
+}
+
+HAL_StatusTypeDef Io_SharedSpi_Receive(
+    const struct SharedSpi *const spi,
+    uint8_t *                     rx_data,
+    uint16_t                      rx_size)
+{
+    Io_SharedSpi_SetNssLow(spi);
+    HAL_StatusTypeDef status =
+        HAL_SPI_Receive(spi->hspi, rx_data, rx_size, 100U);
+    Io_SharedSpi_SetNssHigh(spi);
+
+    return status;
+}
+
+HAL_StatusTypeDef Io_SharedSpi_MultipleTransmitWithoutNssToggle(
+    const struct SharedSpi *const spi,
+    uint8_t *                     tx_data,
+    uint16_t                      tx_size,
+    size_t                        num_tx_data_copies)
+{
+    for (size_t i = 0U; i < num_tx_data_copies; i++)
+    {
+        RETURN_IF_STATUS_NOT_OK(
+            HAL_SPI_Transmit(spi->hspi, tx_data, tx_size, 100U), spi->nss_port,
+            spi->nss_pin)
+    }
+
+    return HAL_OK;
+}
```

Comment:
even though you are not toggling between each sequential spi data frame, shouldn't the nss pin be brought low before transmitting the first frame and then brought high after the last?

Reply:
also this for loop would not work you are not making use of the counter variable in you HAL_SPI_Transmit() command

```
for (size_t i = 0U; i < num_tx_data_copies; i++)
    {
        RETURN_IF_STATUS_NOT_OK(
            HAL_SPI_Transmit(spi->hspi, tx_data + i, tx_size, 100U), spi->nss_port,
            spi->nss_pin)
    }

```
or simply you can remove the for loop altogether:
  HAL_SPI_Transmit(spi->hspi, tx_data, num_tx_data_copies, 100U), spi->nss_port,
            spi->nss_pin)

Reply:
`tx_size` is the size of the `tx_data` buffer you are trying to send. This function is used to send the same `tx_data` multiple times. This ends up sending the `tx_buffer[tx_size]` `num_tx_data_copies` times

Reply:
I specified specifically in the name that we won't be toggling the NSS pin. Here's an example of where this function would be used. 

https://github.com/pasitlaothamatas/Consolidated-Firmware-1/blob/bb58d43bf417f6836aef7603af261d82d800c515/boards/BMS/Src/Io/Io_LTC6813.c#L271

Reply:
Cases like this are IC specific. You probably won't see data being sent like this to other SPI chips. Since we're writing a shared SPI driver I thought it was a better idea to include code that the user can use to create custom functions for sending out packets to a SPI chip.

![image](https://user-images.githubusercontent.com/32230980/94372739-1ec15780-00b5-11eb-9b49-88cfb59e8a7d.png)

Reply:
I'm fine with having this in shared

Reply:
Hmm even if other IC's don't use it? I know the efuse chip doesn't exactly follow this pattern when its communicating with other devices right